src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1207 @@
+    host4_bufs[i] = boost::compute::buffer(ctx_, num_data_ * sizeof(Feature4), 
+                    boost::compute::memory_object::read_write | boost::compute::memory_object::use_host_ptr, 
+                    host4_vecs[i]);
+    host4_ptrs[i] = (Feature4*)queue_.enqueue_map_buffer(host4_bufs[i], boost::compute::command_queue::map_write_invalidate_region,
+                    0, num_data_ * sizeof(Feature4));
+  }
+  #pragma omp parallel for schedule(static)
+  for (unsigned int i = 0; i < dense_feature_map_.size() / dword_features_; ++i) {
+    int tid = omp_get_thread_num();
+    Feature4* host4 = host4_ptrs[tid];
+    auto dense_ind = dense_feature_map_.begin() + i * dword_features_;
+    auto dev_bin_mult = device_bin_mults_.begin() + i * dword_features_;
+    #if GPU_DEBUG >= 1
+    printf("Copying feature ");
+    for (int l = 0; l < dword_features_; ++l) {
+      printf("%d ", dense_ind[l]);
+    }
+    printf("to devices\n");
+    #endif
+    if (dword_features_ == 8) {
+      // one feature datapoint is 4 bits
+      BinIterator* bin_iters[8];
+      for (int s_idx = 0; s_idx < 8; ++s_idx) {
+        bin_iters[s_idx] = train_data_->FeatureIterator(dense_ind[s_idx]);
+        if (dynamic_cast<Dense4bitsBinIterator*>(bin_iters[s_idx]) == 0) {
+          Log::Fatal("GPU tree learner assumes that all bins are Dense4bitsBin when num_bin <= 16, but feature %d is not.", dense_ind[s_idx]);
+        }
+      }
+      // this guarantees that the RawGet() function is inlined, rather than using virtual function dispatching
+      Dense4bitsBinIterator iters[8] = {
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[0]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[1]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[2]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[3]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[4]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[5]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[6]),
+        *static_cast<Dense4bitsBinIterator*>(bin_iters[7])};
+      for (int j = 0; j < num_data_; ++j) {
+        host4[j].s0 = (iters[0].RawGet(j) * dev_bin_mult[0] + ((j+0) & (dev_bin_mult[0] - 1))) 
+                    |((iters[1].RawGet(j) * dev_bin_mult[1] + ((j+1) & (dev_bin_mult[1] - 1))) << 4);
+        host4[j].s1 = (iters[2].RawGet(j) * dev_bin_mult[2] + ((j+2) & (dev_bin_mult[2] - 1))) 
+                    |((iters[3].RawGet(j) * dev_bin_mult[3] + ((j+3) & (dev_bin_mult[3] - 1))) << 4);
+        host4[j].s2 = (iters[4].RawGet(j) * dev_bin_mult[4] + ((j+4) & (dev_bin_mult[4] - 1))) 
+                    |((iters[5].RawGet(j) * dev_bin_mult[5] + ((j+5) & (dev_bin_mult[5] - 1))) << 4);
+        host4[j].s3 = (iters[6].RawGet(j) * dev_bin_mult[6] + ((j+6) & (dev_bin_mult[6] - 1))) 
+                    |((iters[7].RawGet(j) * dev_bin_mult[7] + ((j+7) & (dev_bin_mult[7] - 1))) << 4);
```

Comment:
This part of code makes 4 (or 8) features to a 32-bit feature tuple. So 4 (or 8) features are bundled to one here; a 32-bit host4[i] consists of 4 (or 8) features. CopyTo() seems difficult here, as the destination memory is not linear. We can probably copy each feature data to a temporary array, and use these arrays to build the feature bundles, but this actually has more overhead if these RawGet() can be inlined.

Performance-wise, these RawGet()'s should be inlined by compiler (I checked gcc generated assembly and it is very efficient, but not quite sure on Windows), and it is equivalent to accessing the data_ array of BinIterator using the given index. The tricky part is to make the RawGet() function call non-virtual, as you can see in the code; otherwise it can be really slow.